library/std/src/panicking.rs

```diff
@@ -241,33 +243,53 @@ pub mod panic_count {
     // thread, if that thread currently views `GLOBAL_PANIC_COUNT` as being zero,
     // then `LOCAL_PANIC_COUNT` in that thread is zero. This invariant holds before
     // and after increase and decrease, but not necessarily during their execution.
+    //
+    // Addiitonally, the top bit of GLOBAL_PANIC_COUNT (GLOBAL_ALWAYS_ABORT_FLAG)
+    // records whether panic::always_abort() has been called.  This can only be
+    // set, never cleared.
+    //
+    // This could be viewed as a struct containing a single bit and an n-1-bit
+    // value, but if we wrote it like that it would be more than a single word,
+    // and even a newtype around usize would be clumsy because we need atomics.
+    // But we use such a tuple for the return type of increase().
+    //
+    // Stealing a bit is fine because it just amounts to assuming that each
+    // panicking thread consumes at least 2 bytes of address space.
     static GLOBAL_PANIC_COUNT: AtomicUsize = AtomicUsize::new(0);
 
-    pub fn increase() -> usize {
-        GLOBAL_PANIC_COUNT.fetch_add(1, Ordering::Relaxed);
-        LOCAL_PANIC_COUNT.with(|c| {
-            let next = c.get() + 1;
-            c.set(next);
-            next
-        })
+    pub fn increase() -> (bool, usize) {
+        (
+            GLOBAL_PANIC_COUNT.fetch_add(1, Ordering::Acquire) & ALWAYS_ABORT_FLAG != 0,
```

Comment:
`Ordering::Relaxed` should suffice here (and below in `set_always_abort`). We're not using this flag to synchronize any other data. Any 'happens-before' relationship that might be necessary should already be provided by whatever caused that need. E.g. thread spawning, changing some global allocator property, etc:

If a thread does something like `always_abort(); some_global = 1` and another thread runs code that would somehow be unsound if `some_global == 1` is observed while it doesn't yet see the always-abort flag, that's a problem with the memory ordering on the store+load of `some_global`, and is not solved by an acquire+release ordering on the always-abort flag.

(Specifically, in `always_abort(); thread::spawn(|| ..);`, the always-abort flag is definitely directly visible inside that thread, because spawning a thread synchronizes and guarantees that everything that happened before also already happened from the perspective the new thread as well.)

The other way around, it should never be a problem if `some_global = 1; always_abort();` is executed by one thread and another thread observes the always-abort flag already set while still seeing `some_global != 1`. It might mean that it aborts instead of unrolls when it wasn't strictly necessary yet.